syntax = "proto3";

package filesync;

service FileSyncService {
  // Client -> Server: Upload a file (chunked streaming)
  rpc UploadFile(stream FileChunk) returns (UploadResponse);

  // Client -> Server: Download a file (chunked streaming)
  rpc DownloadFile(FileRequest) returns (stream FileChunk);

  // Server -> Server: Propagate metadata updates
  rpc PropagateMeta(MetadataUpdate) returns (PropagateResponse);

  // Server -> Server: Fetch a specific shard for reconstruction
  rpc FetchShard(ShardRequest) returns (ShardData);

  // Server -> Server: Heartbeat for failure detection
  rpc Heartbeat(HeartbeatRequest) returns (HeartbeatResponse);
}

message FileChunk {
  string file_name = 1;
  string file_hash = 2; // SHA256 of the full file (for verification)
  int32 chunk_index = 3;
  bytes data = 4;
  bool is_last_chunk = 5;
  int64 total_size = 6; // Sent in the first chunk
}

message UploadResponse {
  bool success = 1;
  string message = 2;
  string file_id = 3;
}

message FileRequest {
  string file_name = 1;
}

message MetadataUpdate {
  string file_name = 1;
  int64 version = 2;
  string file_hash = 3;
  int64 timestamp = 4;
  bool is_deleted = 5;
  // Erasure coding metadata
  int32 k_shards = 6;
  int32 m_shards = 7;
  map<int32, string> shard_locations = 8; // shard_index -> node_address
}

message PropagateResponse {
  bool success = 1;
}

message ShardRequest {
  string file_name = 1;
  int32 shard_index = 2;
}

message ShardData {
  bool found = 1;
  bytes data = 2;
}

message HeartbeatRequest {
  string node_id = 1;
  int64 timestamp = 2;
}

message HeartbeatResponse {
  bool alive = 1;
}
